Explain how to paginate with the Analytics Engine SQL API (#6206)
cloudflare/cloudflare-docs: content/analytics/analytics-engine/sql-api.md

Additionally, the [QUANTILEWEIGHTED function](../sql-reference/#quantileweighted) is designed to be used with sample interval as the third argument.

## Pagination

When you have a lot of data stored, the SQL API may not be able to retrieve all the data with one `SELECT` query. Instead you will need to retrieve it using multiple queries.

Traditionally the SQL `OFFSET` feature has been used for pagination, but we do not support it because it is inefficient with big datasets. Instead you can use the timestamp of each row as a cursor.

An example is where you want to retrieve all of your data from a dataset: `SELECT * FROM your_dataset`. If you only have a few thousand rows then that query should execute fine, but if you have millions of rows then it is likely to fail. Instead you can break down the query into multiple sequential queries.

> First run your desired query, adding `ORDER BY timestamp DESC` and a sensible `LIMIT`:

Reviewer: Can we offer any guidance on what a sensible value might be?
What factors affect this?

Author: I think I'm going to suggest `1000` items as a sensible limit. Queries are expensive to run and if we suggest something like `100` it'll add up a lot more.

```

> > SELECT * FROM your_dataset ORDER BY timestamp DESC LIMIT 10000

Reviewer: I think it might be worth breaking each of these examples into two code blocks with appropriate syntax highlighting on each.

{
    "data": [
        [...],
        {
            "timestamp": "2022-10-10 13:56:48",
            [...]
        }
    ]
}
```

Then run a second query, using the timestamp of the final row in the `WHERE` clause of the next query:

```

> > SELECT * FROM your_dataset WHERE timestamp < toDateTime("2022-10-10 13:56:48") ORDER BY timestamp DESC LIMIT 1000

Reviewer: ```suggestion
> SELECT * FROM your_dataset WHERE timestamp < toDateTime("2022-10-10 13:56:48") ORDER BY timestamp DESC LIMIT 10000
```

Additionally, the [QUANTILEWEIGHTED function](../sql-reference/#quantileweighted) is designed to be used with sample interval as the third argument.

## Pagination

When you have a lot of data stored, the SQL API may not be able to retrieve all the data with one `SELECT` query. Instead you will need to retrieve it using multiple queries. We suggest retrieving 1000 records at a time.

> Traditionally the SQL `OFFSET` feature has been used for pagination, but we do not support it because it is inefficient with big datasets. Instead you can use the timestamp of each row as a cursor.

Reviewer 2: ```suggestion
Traditionally, the SQL `OFFSET` feature has been used for pagination, but we do not support it because it is inefficient with big datasets. Instead you can use the timestamp of each row as a cursor.
```

> An example is where you want to retrieve all of your data from a dataset: `SELECT * FROM your_dataset`. If you only have a few thousand rows then that query should execute fine, but if you have millions of rows then it is likely to fail. Instead you can break down the query into multiple sequential queries.

Reviewer: Few thousand seems at odds with the recommendation of 1000.

> An example is where you want to retrieve all of your data from a dataset: `SELECT * FROM your_dataset`. If you only have a few thousand rows then that query should execute fine, but if you have millions of rows then it is likely to fail. Instead you can break down the query into multiple sequential queries.

Reviewer 2: ```suggestion
An example might be when you want to retrieve all of your data from a dataset: `SELECT * FROM your_dataset`. If you only have a few thousand rows then that query should execute fine, but if you have millions of rows then it is likely to fail. Instead you can break down the query into multiple sequential queries.
```

> First run your desired query, adding `ORDER BY timestamp DESC` and our recommended `LIMIT` of 1000:

Reviewer 2: ```suggestion
First, run your desired query, adding `ORDER BY timestamp DESC` and our recommended `LIMIT` of 1000:
```

```sql
SELECT * FROM your_dataset ORDER BY timestamp DESC LIMIT 1000
```

The response will include up to 1000 rows. Since the rows are ordered by descending timestamp, the oldest row has the oldest timestamp:

```json
{
    "data": [
        [...],
        {
            "timestamp": "2022-10-10 13:56:48",
            [...]
        }
    ]
}
```

> Next run a second query, using the timestamp of the final row in the `WHERE` clause of the next query:

Reviewer 2: ```suggestion
Next, run a second query, using the timestamp of the final row in the `WHERE` clause of the next query:
```

```sql
SELECT * FROM your_dataset WHERE timestamp < toDateTime("2022-10-10 13:56:48") ORDER BY timestamp DESC LIMIT 1000
```

> You'll get the next 1000 older rows:

Reviewer 2: ```suggestion
You will get the next 1000 older rows:
```

> You will get the next 1000 older rows:

Reviewer 3: What if the next batch will get 1000 rows with the same timestamp?

I think we should highlight here an assumption that it works for a scenario when we have at most 1 request per second.

Would also be nice to notice that we SQL API does not support `LIMIT n OFFSET m`.
